### PR TITLE
Blueprint contents localisation

### PR DESCRIPTION
**Pull request by gundelsby**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
i18n data contained in blueprint bundles is added to the existing translator function in core, making it possible to have localised content in blueprint code

* **Other information**:
Changes in the blueprints project:
https://github.com/nrkno/tv-automation-sofie-blueprints/pull/65

**Status**
- [X] The functionality has been tested by the PR author
- [ ] The functionality has been tested by NRK

### CODE REVIEW

**Comment by gundelsby** [comment]
Closed due to branching out from the wrong parent. New PR here: https://github.com/nrkno/tv-automation-server-core/pull/287

File: meteor/server/api/ingest/bucketAdlibs.ts
Comment (jstarpl) [inline meteor/server/api/ingest/bucketAdlibs.ts:36]: What's this about? You can use `unprotectStringArray` if you need to pass some protectedStrings array as a string array.

**Review by jstarpl — CHANGES_REQUESTED** [review]
I think it might be worth to convert the subscription/reactive query to a periodic method call, as this might be a pretty substantial subscription (in terms of memory size).

File: meteor/client/ui/i18n.ts
Comment (gundelsby) [reply meteor/client/ui/i18n.ts:59]: Yeah, that's fair critique. This change has been put on hold until after the performance sprint anyway, but I don't think we should release this until we've had a proper look at this either.

File: meteor/server/api/blueprints/api.ts
Comment (gundelsby) [reply meteor/server/api/blueprints/api.ts:162]: It's only added to the StudioBlueprintManifest and ShowStyleBlueprintManifest, not the SystemBlueprintManifest. SomeBlueprintManifest  can be either of those. The changes are in a nightly version which is built for this branch like you suggested.

We could keep it as is, I could explicitly check for those types, or we could add it to the Some-type. I don't really feel strongly about any of the options.

File: meteor/server/api/ingest/bucketAdlibs.ts
Comment (gundelsby) [reply meteor/server/api/ingest/bucketAdlibs.ts:36]: That's a good tip, thanks. I'll push an update 👍

Comment (jstarpl) [reply meteor/server/api/ingest/bucketAdlibs.ts]: I was just hoping that there's a way of not running the map and filter. I don't think either `blueprintId` or `studio.blueprintId` can be undefined (as per typings), so just a simple overload should be fine.

Comment (Julusian) [reply meteor/server/api/ingest/bucketAdlibs.ts]: this has been changed in #336, which will be merged in at the same time, so isnt worth worrying about

File: meteor/server/api/blueprints/api.ts
Comment (jstarpl) [reply meteor/server/api/blueprints/api.ts:194]: Let's have @nytamin decide, he's the "strict typings" ombudsman.

File: meteor/client/ui/i18n.ts
Comment (nytamin) [reply meteor/client/ui/i18n.ts:59]: I agree, since this is going to be basically static data, a method call will be much lighter than setting up a subscription

File: meteor/lib/api/rundown.ts
Comment (nytamin) [inline meteor/lib/api/rundown.ts:86]: typo

File: meteor/server/api/translationsBundles.ts
Comment (nytamin) [inline meteor/server/api/translationsBundles.ts:42]: This only creates an implicit (&loose) connection between the namespace-language pair and the _id. I would very much prefer that to be an explicit one (which also should include the _id of the blueprint that created it, to ensure uniqueness).
Example: `return getHash('${blueprint}_${namespace}_${language}')`.

File: meteor/lib/collections/TranslationsBundles.ts
Comment (nytamin) [inline meteor/lib/collections/TranslationsBundles.ts:11]: I'm missing a `blueprint: BlueprintId` here, so that a bundle is explicitly tied to the blueprint that created it.

File: meteor/server/api/blueprints/api.ts
Comment (nytamin) [reply meteor/server/api/blueprints/api.ts:194]: I'm in favor of the stricter types solution, as that decreases the risk of future bugs during refactoring.

File: meteor/server/api/playout/playout.ts
Comment (nytamin) [inline meteor/server/api/playout/playout.ts:1217]: todo

File: meteor/server/api/translationsBundles.ts
Comment (nytamin) [inline meteor/server/api/translationsBundles.ts:15]: So I'm confused about the "namespace". If it is equal to the _id of the blueprint that created it, I think it should be named `blueprintId`

File: meteor/server/api/translationsBundles.ts
Comment (nytamin) [inline meteor/server/api/translationsBundles.ts:36]: This is just for debugging during dev, and should be removed before merging, right?

File: meteor/server/api/translationsBundles.ts
Comment (nytamin) [inline meteor/server/api/translationsBundles.ts:22]: Is there a reason for using the callback pattern here?
If there is an error it wont be displayed to the user, which might be unhelpful.

File: meteor/server/publications/translationsBundles.ts
Comment (nytamin) [inline meteor/server/publications/translationsBundles.ts:8]: to be removed, if we use a method instead

**Review by nytamin — CHANGES_REQUESTED** [review]
A few notes, most of which I have talked to @gundelsby about before (but to keep track I put them in writing now as well).

File: meteor/server/api/translationsBundles.ts
Comment (gundelsby) [reply meteor/server/api/translationsBundles.ts:36]: Yup

**Comment by gundelsby** [comment]
This is now working, so I've converted the PR from a draft and will start working on the QA taking into consideration the previous comments from @jstarpl and @nytamin that are still valid.

**Review by jstarpl — APPROVED** [review]

File: meteor/client/ui/i18n.ts
Comment (nytamin) [reply meteor/client/ui/i18n.ts:96]: I'm bumping this up again, as I still think it would be better to use a method than a subscription, since the data is going to be static.

File: meteor/lib/api/userActions.ts
Comment (nytamin) [inline meteor/lib/api/userActions.ts:164]: I don't think it's a good idea to use the UserActions API for fetching the translations, as that is tied to user-facing things (since it logs an entry to the userActions-log). For this I would instead just expose a plain Meteor method. It's not as type-safe, but it is what we have.

File: meteor/server/api/translationsBundles.ts
Comment (nytamin) [inline meteor/server/api/translationsBundles.ts:31]: Question: Is there a specific reason for the lookup with `getExistingId`? Why not just use createBundleId right away? (that would be more performant). If there is a reason, plz add it as a comment in the code.

File: meteor/server/api/blueprints/context/context.ts
Comment (nytamin) [inline meteor/server/api/blueprints/context/context.ts:165]: Should this temporary feature be removed?

File: meteor/server/api/userActions.ts
Comment (nytamin) [inline meteor/server/api/userActions.ts:1069]: See my other comment; I think this should be a MeteorMethod, not a UserAction.

**Review by nytamin — CHANGES_REQUESTED** [review]

File: meteor/server/api/blueprints/context/context.ts
Comment (Julusian) [reply meteor/server/api/blueprints/context/context.ts:165]: we dont have a way of getting these to the user currently, but they really should be.
so this is a hack to make them go somewhere (rather than into the void) until a follow up task is written and done

Comment (Julusian) [reply meteor/server/api/blueprints/context/context.ts:165]: but I am thinking I will end up making the blueprints do a log* everytime they do a notifyUser* so that we have a record of the note with more technical details, so maybe this isnt necessary?

File: meteor/lib/api/userActions.ts
Comment (Julusian) [reply meteor/lib/api/userActions.ts:164]: I agree, the purpose of the UserActions API is to provide a record of what the user did, and this does not want to be logged like that

**Comment by jstarpl** [comment]
@gundelsby Looks good to me!

**Pull request merged**
